src/main/java/uk/gov/hmcts/reform/idam/client/IdamClient.java

```diff
@@ -37,17 +45,38 @@
     private IdamApi idamApi;
     private OAuth2Configuration oauth2Configuration;
 
+    private static Cache<String, TokenResponse> accessTokenCache;
+    private final long refreshTokenBeforeExpiry;
+    private final boolean  enableCache;
+
     @Autowired
-    public IdamClient(IdamApi idamApi, OAuth2Configuration oauth2Configuration) {
+    public IdamClient(IdamApi idamApi, OAuth2Configuration oauth2Configuration,
+        @Value("${idam.client.enable-cache:false}") boolean enableCache,
+        @Value("${idam.client.refresh-before-expire-in-sec:120}") long refreshTokenBeforeExpiry) {
         this.idamApi = idamApi;
         this.oauth2Configuration = oauth2Configuration;
+        this.enableCache = enableCache;
+        this.refreshTokenBeforeExpiry = refreshTokenBeforeExpiry;
+        this.accessTokenCache = Caffeine.newBuilder()
+            .expireAfter(new AccessTokenCacheExpiry())
+            .build();
     }
 
     public UserDetails getUserDetails(String bearerToken) {
         return idamApi.retrieveUserDetails(bearerToken);
     }
 
     public String getAccessToken(String username, String password) {
+        if (enableCache) {
+            return
+                accessTokenCache
+                    .get(username, b -> retrieveAccessToken(username, password)).accessToken;
+        } else {
+            return retrieveAccessToken(username, password).accessToken;
+        }
```

Comment:
I guess nicer way would be some sort of in-build `BiFunction` which consumes username and password and returns `accessToken`. But that's just personal too functional mind chipping in here

Reply:
I didn't get it. 
I need TokenResponse in cache to calculate expiry, so cache returns TokenRespone